### ISSUE TITLE
Future interruption (raise, cancel) doesn't propagate through `collect`

### ISSUE THREAD

**Issue by tudor** [facebook/folly#1237]
One would expect that interrupting the future resulting from `collect` would interrupt all the source futures, but this doesn't happen.

This prints 0, not 2:

```
    int n = 0;

    auto [p1, f1] = folly::makePromiseContract<folly::Unit>();
    p1.setInterruptHandler([&](const folly::exception_wrapper& ew) { ++n; });

    auto [p2, f2] = folly::makePromiseContract<folly::Unit>();
    p2.setInterruptHandler([&](const folly::exception_wrapper& ew) { ++n; });

    auto f = folly::collect(std::move(f1), std::move(f2));

    f.cancel();

    printf("%d\n", n);
```

Is there a better way to accomplish this?

**Comment by yfeldblum** [comment]
This is true - cancellation is only partially supported in folly futures. It's integrated into `.then`-family of functions, but not into all the combinators.

In contrast to coroutines with `folly::coro::Task`, where cancellation is designed differently and integrated better including through the coroutine `collect` functions.

**Comment by tudor** [comment]
Are there any plans to fix this, or is `folly::Future` essentially deprecated, with coroutines being The Future? (Switching my code to an experimental version of clang is not exactly an option for me, sadly.)

**Comment by yfeldblum** [comment]
I did briefly attempt this at one point, but the way cancellation is done in folly futures is super finicky and I didn't figure out all the test failures.

More of the active work, currently, is on folly coroutines support. We want to be prepared when C++20 ships and g++ and clang++ are released with full support for language-level coroutines. As an async programming paradigm, coroutines is overall much safer.

I don't have a great answer. If this is for internal code, consider writing a custom variant `collectWithInterrupt` that propagates cancellation or doing something else around `folly::collect` using `folly::CancellationToken`.
